datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,336 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Distribution partitioner class that works with Hugging Face Datasets."""
+
+
+from collections import Counter
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class DistributionPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on a distribution.
+
+    Inspired from implementations of Li et al.
+    Federated Optimization in Heterogeneous Networks (2020)
+    https://arxiv.org/abs/1812.06127.
+
+    Parameters
+    ----------
+    distribution_array : Union[NDArrayInt, NDArrayFloat]
+        A 2-dimensional numpy array of the probability distribution of samples
+        for all labels in all partitions. The array shape should be
+        (`num_unique_labels`,
+        `num_unique_labels_per_partition*num_partitions/num_unique_labels`),
+        such that the first row of the array corresponds to the sample distribution
+        of the first unique label (in ascending order). The values may be scaled per
+        label such that the sum of the label distributions across all partitions are
+        equal (or close to) to the original unpartitioned label distribution
+        - see the `rescale` argument.
+    num_partitions : int
+        The total number of partitions that the data will be divided into.
+    num_unique_labels_per_partition : int
+        Number of unique labels assigned to a single partition.
+    partition_by : str
+        Column name of the labels (targets) based on which sampling works.
+    preassigned_num_samples_per_label : int
+        The minimum number of samples that each label in each partition will have.
+    rescale : bool, default=True
+        Whether to partition samples according to the values in
+        `distribution_array` or rescale based on the original unpartitioned
+        class label distribution. `float` values are rounded to the nearest `int`.
```

Comment:
Related to this argument - we are not exhausting the samples. The purpose of rescale is (at least in my mind, to use all the samples). If not enabled by default, we should add an argument to allow the use of all samples - counter the float to int rounding.

Reply:
Good point, I've added the exhaustive rescaling. For now, it is set to be exhaustive when `rescale = True`. I think it makes sense. I've also updated the PDF above and it's similar, which is good.